extras/kms/data_key.go

```diff
@@ -0,0 +1,84 @@
+package kms
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/hashicorp/go-dbw"
+)
+
+// dataKey represents the DEKs (keys to encrypt data) of the system and must
+// have a parent root key and a purpose.
+type dataKey struct {
+	// PrivateId is used to access the key
+	PrivateId string `json:"private_id,omitempty" gorm:"primary_key"`
+	// RootKeyId for the key
+	RootKeyId string `json:"root_key_id,omitempty" gorm:"default:null"`
+	// Purpose of the the key
+	Purpose KeyPurpose `json:"purpose,omitempty" gorm:"default:null"`
+	// CreateTime from the RDBMS
+	CreateTime time.Time `json:"create_time,omitempty" gorm:"default:current_timestamp"`
+}
+
+// newDataKey creates a new in memory data key.  This key is used for wrapper
+// operations.  No options are currently supported.
+func newDataKey(rootKeyId string, purpose KeyPurpose, _ ...Option) (*dataKey, error) {
+	const op = "kms.NewDataKey"
+	if rootKeyId == "" {
+		return nil, fmt.Errorf("%s: missing root key id: %w", op, ErrInvalidParameter)
+	}
+	switch purpose {
+	case KeyPurposeUnknown:
+		return nil, fmt.Errorf("%s: missing purpose: %w", op, ErrInvalidParameter)
+	case KeyPurposeRootKey:
+		return nil, fmt.Errorf("%s: cannot be a purpose of %q: %w", op, purpose, ErrInvalidParameter)
+	}
+	c := &dataKey{
+		RootKeyId: rootKeyId,
+		Purpose:   purpose,
+	}
+	return c, nil
+}
+
+// Clone creates a clone of the DataKey
+func (k *dataKey) Clone() *dataKey {
```

Comment:
I'm curious why you didn't go with proto for types, given that most of the values in go-kms-wrapping are protos to ensure compatibility outside of Go. I don't have a problem with it, I will just note that one benefit is that it's impossible to forget to add later values to Clone :-D

Reply:
Actually if you wanted that safety you could do the same thing with `json` marshaling.

Reply:
I didn't go with protos because we didn't need them (including in boundary).  I didn't go with marshaling (gob or json) because it's seems like a false sense of security for future changes since those marshaling techniques only work if all fields are public.